### ISSUE TITLE
Bug - Set Filter reset between renders

### ISSUE THREAD

**Issue by definitelycarter** [ag-grid/ag-grid#7876]
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => see 'Providing a Reproducible Scenario'
[] feature request => do not use Github for feature requests, see 'Customers of AG Grid'
[] support request => see 'Requesting Community Support'
```

**Providing a Reproducible Scenario**

Codesandbox: [link](https://codesandbox.io/p/sandbox/ag-grid-set-filter-bug-with-client-mode-ctzfm3)

**Current behavior (steps to reproduce)**

1. Open the above codesandbox link.
2. Filter the column using the configured set filter.
3. Once rows are filtered, click the button on the grid to trigger a re-render.

**Expected Behavior** - The filter for the column should be persisted between re-renders.
**Actual Behavior** - The filter is reset without any filters applied.
**Note** The workaround is to uncomment cellDataType on line 32 in App.js or to downgrade to ag grid v30.x

**My Debugging**

I found that in this particular scenario, when the `cellDataType` is not set to false the following happens:

1. The data type for the column is inferred to an 'object' [here](https://github.com/ag-grid/ag-grid/blob/262783e757122a1c3fefb1f37f65d0bdd4dc2078/community-modules/core/src/columns/dataTypeService.ts#L697) and the `keyCreator`, `comparator` and other fields are set (despite the column definition already having these properties set).
2. The Set Filter has logic [here](https://github.com/ag-grid/ag-grid/blob/7c6d881ee9e437003aaecdc9847b2b598a1f8c5f/enterprise-modules/set-filter/src/setFilter/setFilter.ts#L168) and [here](https://github.com/ag-grid/ag-grid/blob/7c6d881ee9e437003aaecdc9847b2b598a1f8c5f/enterprise-modules/set-filter/src/setFilter/setFilter.ts#L200) that determines _important_ fields that when changed trigger a forced reload.

**In Summary** The user has specified a custom `keyCreator`, `comparator` and other field(s) that do not change between renders; However, because the data type is inferred to an `object`, the  `keyCreator`, `comparator`, etc. are reset to a brand new function and thereby failing an equality comparison and triggering a reset of the filter state. In my opinion this is a regression as I am unable to reproduce this scenario on v30.x and this is new behavior without any new end user change(s). Cell Data Type inference should probably default to false (instead of true) or at least avoid setting those important properties if they are already set on the column def.

* **AG Grid version:** 31.2.0 (bug also exists on latest)

* **Browser:** any browser

* **Language:** [all]

**Comment by AG-Zoheil** [comment]
Hello 

Thank you for reporting this - much appreciated!

Based on your description we've now reproduced this issue as a bug.

We have added this bug to our backlog and we are tracking it with the following reference: **AG-11413**

We try to fix bugs from one release to the next, so this should be fixed in the next release or the one after if it was raised too close to the next release date.

See whether this item will be in the next release by checking the NEXT RELEASE checkbox on the product pipeline page:
https://www.ag-grid.com/ag-grid-pipeline/

The best way to track this is to sign up for AG Grid new release notifications using [the instructions here](https://ag-grid.zendesk.com/hc/en-us/articles/360002648058-How-do-I-sign-up-for-ag-Grid-release-notifications-). This way you'll know as soon as a new version is out and you can check whether this specific item was implemented on [the changelog page](https://www.ag-grid.com/ag-grid-changelog/).

In the meantime as a workaround you can provide the `keyCreator` directly on the column definition instead of the `filterParams` which should work.

Thanks again for bringing this up with us.

Kind regards,
Zoheil

**Comment by fkul** [comment]
Hello,

Thank you for addressing this. Unfortunately I'm still experiencing that bug in v32.1.0. You can find the updated CodeSandbox here:

https://codesandbox.io/p/sandbox/ag-grid-set-filter-bug-with-client-mode-forked-vrmc64

The only change I made, besides upgrading to v32.1.0, was moving the `keyCreator` into the column definitions.

I've started having this issue after upgrading from v31.0.3 to v32.1.0, so I would still consider this a regression.
